proxy/envoy/config.go

```diff
@@ -303,9 +303,11 @@ func buildInboundFilters(instances []*model.ServiceInstance) (HTTPRouteConfigs,
 			http.VirtualHosts = append(http.VirtualHosts, host)
 		case model.ProtocolTCP:
 			cluster := buildInboundCluster(service.Hostname, endpoint.Port, port.Protocol)
-			route := buildTCPRoute(cluster, endpoint.Address, endpoint.Port)
 			config := tcpConfigs.EnsurePort(port.Port)
-			config.Routes = append(config.Routes, route)
+			config.Routes = append(config.Routes,
+				buildTCPRoute(cluster, []string{endpoint.Address}, endpoint.Port),
+				buildTCPRoute(cluster, []string{service.Address, endpoint.Address}, port.Port),
```

Comment:
Listener port for inbound TCP connections is always `targetPort`, not the service port, since port rewriting happens in the node network namespace. In HTTP protocol, the host header still keeps the old service port. Take a look here: https://github.com/istio/manager/blob/master/proxy/envoy/config.go#L302. For example, consider a service A with port 80 and target port 8080, we'll generate an entry for listener port 8080 and HTTP route with virtual host "A:80". Does that make sense? I can explain more if it's still confusing.